cluster-api/pkg/controller/machine/controller.go

```diff
@@ -79,37 +79,46 @@ func (c *MachineControllerImpl) Init(arguments sharedinformers.ControllerInitArg
 // Reconcile handles enqueued messages. The delete will be handled by finalizer.
 func (c *MachineControllerImpl) Reconcile(machine *clusterv1.Machine) error {
 	// Implement controller logic here
-	glog.Infof("Running reconcile Machine for %s\n", machine.Name)
-	exist, err := c.actuator.Exists(machine)
-	if err == nil {
-		if !exist {
-			// Machine resource created. Machine does not yet exist.
-			glog.Infof("reconciling machine object %v triggers idempotent create.", machine.ObjectMeta.Name)
-			err = c.create(machine)
-		} else {
-
-			if !machine.ObjectMeta.DeletionTimestamp.IsZero() {
-				// no-op if finalizer has been removed.
-				if !util.Contains(machine.ObjectMeta.Finalizers, clusterv1.MachineFinalizer) {
-					glog.Infof("reconciling machine object %v causes a no-op as there is no finalizer.", machine.ObjectMeta.Name)
-					return nil
-				}
-				if cfg.ControllerConfig.InCluster && util.IsMaster(machine) {
-					glog.Infof("skipping reconciling master machine object %v", machine.ObjectMeta.Name)
-				} else {
-					glog.Infof("reconciling machine object %v triggers delete.", machine.ObjectMeta.Name)
-					err = c.delete(machine)
-				}
-			} else {
-				glog.Infof("reconciling machine object %v triggers idempotent update.", machine.ObjectMeta.Name)
-				err = c.update(machine)
-			}
+	name := machine.Name
+	glog.Infof("Running reconcile Machine for %s\n", name)
+	isDeleting := !machine.ObjectMeta.DeletionTimestamp.IsZero()
+	hasMachineFinalizer := util.Contains(machine.ObjectMeta.Finalizers, clusterv1.MachineFinalizer)
+
+	if isDeleting {
+		// no-op if finalizer has been removed.
+		if !hasMachineFinalizer {
+			glog.Infof("reconciling machine object %v causes a no-op as there is no finalizer.", name)
+			return nil
+		}
+		if cfg.ControllerConfig.InCluster && util.IsMaster(machine) {
+			glog.Infof("skipping reconciling master machine object %v", name)
+			return nil
+		}
+		glog.Infof("reconciling machine object %v triggers delete.", name)
+		if err := c.delete(machine); err != nil {
+			return err
+		}
+		// Remove finalizer on successful deletion.
+		glog.Infof("machine object %v deletion successful, removing finalizer.", name)
+		machine.ObjectMeta.Finalizers = util.Filter(machine.ObjectMeta.Finalizers, clusterv1.MachineFinalizer)
+		if _, err := c.machineClient.Update(machine); err != nil {
+			return err
```

Comment:
Would want to rewrap this error otherwise when the error comes out the other end, it will not be clear that this was an update to remove the finalizer that failed.

Reply:
That and/or have a log statement.